### PR TITLE
drop pytz dependency

### PR DESCRIPTION
**Pull request by varoonp123**
pytz docs mention that it shouldnt be required anymore and it is always good to cut back on unneeded 3rd party dependencies

## Summary
Make pytz only required prior to the introduction of zoneinfo in the standard library  #714 

## Checklist
Delete items not relevant to your PR:
- [x] Unit and integration tests covering the common scenarios were added (Existing tests pass locally. If we agree with the basic idea/implementation, I will add more tests for some edge cases)
- [x] A human-readable description of the changes was provided to include in CHANGELOG

### CODE REVIEW

File: clickhouse_connect/cc_sqlalchemy/datatypes/sqltypes.py [hunk 1]
```diff
@@ -1,7 +1,11 @@
 from collections.abc import Sequence
 from enum import Enum as PyEnum
 
-import pytz
+try:
+    import zoneinfo
+except ImportError:
+    from backports import zoneinfo
```
Comment (varoonp123) [inline clickhouse_connect/cc_sqlalchemy/datatypes/sqltypes.py:7]: The net-addition in the diff comes from doing this over and over again. If we are good fully dropping python < 3.9, then we can drop backports.zoneinfo altogether.

**Comment by varoonp123** [comment]
Ill make some fixes for these Copilot comments tomorrow. 2 questions:
1. If an environment doesn't have an IANA timezone data source, how should we proceed? Maybe make an optional install feature `pip install clickhouse-client[tzdata]` that pulls in the [`tzdata` package maintained by the Python github org](https://github.com/python/tzdata) and give a nice error message suggesting this install feature? Apparently this applies to Windows environments.
2. Are we still "soft" supporting python <=3.8? If we agree to only support python3.9+, then I can drop all of the backports nonsense.

These install features/conditional dependencies are all in service of minimizing dependencies for "up-to-date (debian) linux servers not running EOL versions of CPython" which is my main deployment target/environment and I imagine the most common deployment environment. @joe-clickhouse @peter-leonov-ch

**Comment by joe-clickhouse** [comment]
Hi @varoonp123 thanks for this PR! This has been on my list to do for a while now so it'll be nice to get it done, though it's definitely non-trivial, so I appreciate the time and effort.

To start with your specific questions:
1. yeah, I like the optional idea. But for Windows we should make it included by default so `'tzdata; sys_platform == "win32"'` in the `setup.py` `install_requires`. And then also `"tzdata": ["tzdata"]` in `extras_requires` for non-windows users who hit this unexpectedly (e.g. slim docker images, etc.) with a clear error message to install the optional tzdata via `pip install clickhouse-connect[tzdata]`.
2. Don't worry compat issues with anything under Python 3.10. Moving forward, we require Python>=3.10 so no need to worry about the backport stuff.

As for code specific things, I'll make additional notes.

File: clickhouse_connect/cc_sqlalchemy/datatypes/sqltypes.py (hunk 1 above)
Comment (joe-clickhouse) [reply clickhouse_connect/cc_sqlalchemy/datatypes/sqltypes.py]: As mentioned, we can get rid of all this stuff. We only support Python>=3.10.

**Comment by varoonp123** [comment]
@joe-clickhouse Thanks for the feedback. Ill update based on your and copilot's comments in the next couple of days. I know we have another PR for fixed offset timezones. Those should work nicely with `datetime.timezone` and not require pytz.

File: clickhouse_connect/driver/client.py
Comment (joe-clickhouse) [inline clickhouse_connect/driver/client.py]: Swap this for `ZoneInfoNotFoundError`. Then add a logger warning about installing tzdata.

File: clickhouse_connect/driver/client.py [hunk 2]
```diff
@@ -296,7 +300,7 @@ def _prep_query(self, context: QueryContext):
     def _check_tz_change(self, new_tz) -> tzinfo | None:
         if new_tz:
             try:
-                new_tzinfo = pytz.timezone(new_tz)
+                new_tzinfo = zoneinfo.ZoneInfo(new_tz)
                 if new_tzinfo != self.server_tz:
                     return new_tzinfo
             except UnknownTimeZoneError:
```
Comment (joe-clickhouse) [inline clickhouse_connect/driver/client.py]: Swap this for `ZoneInfoNotFoundError`. Then add a logger warning about installing tzdata.

File: clickhouse_connect/driver/query.py [hunk 3]
```diff
@@ -121,15 +125,15 @@ def __init__(
             raise ProgrammingError(f'tz_mode must be "naive_utc", "aware", or "schema", got "{self.tz_mode}"')
         if isinstance(query_tz, str):
             try:
-                query_tz = pytz.timezone(query_tz)
+                query_tz = zoneinfo.ZoneInfo(query_tz)
             except UnknownTimeZoneError as ex:
```
Comment (joe-clickhouse) [inline clickhouse_connect/driver/query.py]: Same as in client.py, swap this for `ZoneInfoNotFoundError`. Then add a logger warning about installing tzdata.

File: clickhouse_connect/driver/query.py [hunk 4]
```diff
@@ -121,15 +125,15 @@ def __init__(
...
                 raise ProgrammingError(f"query_tz {query_tz} is not recognized") from ex
         self.query_tz = query_tz
         if column_tzs is not None:
             for col_name, timezone in column_tzs.items():
                 if isinstance(timezone, str):
                     try:
-                        timezone = pytz.timezone(timezone)
+                        timezone = zoneinfo.ZoneInfo(timezone)
                         column_tzs[col_name] = timezone
                     except UnknownTimeZoneError as ex:
```
Comment (joe-clickhouse) [inline clickhouse_connect/driver/query.py]: Same as in client.py, swap this for `ZoneInfoNotFoundError`. Then add a logger warning about installing tzdata.

File: clickhouse_connect/driver/query.py [hunk 5]
```diff
@@ -51,7 +55,7 @@ def __init__(
         query_formats: dict[str, str] | None = None,
         column_formats: dict[str, str | dict[str, str]] | None = None,
         encoding: str | None = None,
-        server_tz: tzinfo = pytz.UTC,
+        server_tz: tzinfo = zoneinfo.ZoneInfo("UTC"),
         use_none: bool | None = None,
```
[the remark this answers is not in the document]
Comment (joe-clickhouse) [reply clickhouse_connect/driver/query.py]: I agree with copilot here. Let's switch all UTC sentinels from `zoneinfo.ZoneInfo("UTC")` to `datetime.timezone.utc`. Since we require Python>=3.10 stdlib `zoneinfo` will be available in all interpreters.

File: clickhouse_connect/driver/client.py [hunk 6]
```diff
@@ -296,7 +300,7 @@ def _prep_query(self, context: QueryContext):
     def _check_tz_change(self, new_tz) -> tzinfo | None:
         if new_tz:
             try:
-                new_tzinfo = pytz.timezone(new_tz)
+                new_tzinfo = zoneinfo.ZoneInfo(new_tz)
                 if new_tzinfo != self.server_tz:
```
Comment (joe-clickhouse) [inline clickhouse_connect/driver/client.py:302]: Once we swap out the UTC sentinels, we'll need to guard the tz comparison:

```python
  if tzutil.is_utc_timezone(new_tzinfo) and tzutil.is_utc_timezone(self.server_tz):                                                                                                                            
      return None                                                                                                                                                                                              
  if new_tzinfo != self.server_tz:
      return new_tzinfo
```

File: tests/integration_tests/test_params.py [hunk 7]
```diff
@@ -16,10 +16,10 @@ def test_params(param_client: Client, call, table_context: Callable):
...
         assert "rr" in result.first_item["name"]
 
-    first_date = datetime.strptime("Jun 1 2005  1:33PM", "%b %d %Y %I:%M%p")
-    first_date = param_client.server_tz.localize(first_date)
-    second_date = datetime.strptime("Dec 25 2022  5:00AM", "%b %d %Y %I:%M%p")
-    second_date = param_client.server_tz.localize(second_date)
+    first_date = datetime.strptime("Jun 1 2005  1:33PM", "%b %d %Y %I:%M%p").astimezone(param_client.server_tz)
+    # first_date = param_client.server_tz.localize(first_date)
+    second_date = datetime.strptime("Dec 25 2022  5:00AM", "%b %d %Y %I:%M%p").astimezone(param_client.server_tz)
+    # second_date = param_client.server_tz.localize(second_date)
```
Comment (joe-clickhouse) [inline tests/integration_tests/test_params.py]: Yes agree here. Let's replace `.astimezone(param_client.server_tz)` with `.replace(tzinfo=param_client.server_tz)`. Then let's also delete the commented lines.

File: clickhouse_connect/tools/datagen.py [hunk 8]
```diff
@@ -95,13 +99,13 @@ def random_value_gen(ch_type: ClickHouseType, col_def: RandomValueDef):
     if isinstance(ch_type, FixedString):
         return lambda: bytes(int(random() * 256) for _ in range(ch_type.byte_size))
     if isinstance(ch_type, DateTime):
-        if col_def.server_tz == pytz.UTC:
+        if col_def.server_tz == zoneinfo.ZoneInfo("UTC"):
```
Comment (joe-clickhouse) [inline clickhouse_connect/tools/datagen.py]: We'll need to replace with `if tzutil.is_utc_timezone(col_def.server_tz):`

File: clickhouse_connect/tools/datagen.py [hunk 9]
```diff
@@ -95,13 +99,13 @@ def random_value_gen(ch_type: ClickHouseType, col_def: RandomValueDef):
...
             return random_datetime
         timezone = col_def.server_tz
         return lambda: random_datetime_tz(timezone)
     if isinstance(ch_type, DateTime64):
         prec = ch_type.prec
-        if col_def.server_tz == pytz.UTC:
+        if col_def.server_tz == zoneinfo.ZoneInfo("UTC"):
```
Comment (joe-clickhouse) [inline clickhouse_connect/tools/datagen.py]: We'll need to replace with `if tzutil.is_utc_timezone(col_def.server_tz):`

File: tests/integration_tests/test_timezones.py [hunk 10]
```diff
@@ -45,12 +50,12 @@ def test_server_timezone(param_client: Client, call):
     test_datetime = datetime(2023, 3, 18, 16, 4, 25)
     try:
         date = call(param_client.query, "SELECT toDateTime(%s) as st", parameters=[test_datetime]).first_row[0]
-        if param_client.server_tz == pytz.UTC:
+        if param_client.server_tz == zoneinfo.ZoneInfo("UTC"):
```
Comment (joe-clickhouse) [inline tests/integration_tests/test_timezones.py]: We'll need to replace with `if tzutil.is_utc_timezone(param_client.server_tz):`

File: tests/integration_tests/test_timezones.py [hunk 11]
```diff
@@ -191,8 +196,8 @@ def test_tz_mode(param_client: Client, call):
         query_tz="UTC",
         tz_mode="aware",
     ).first_row
-    assert row[0].tzinfo == pytz.UTC
-    assert row[1].tzinfo == pytz.UTC
+    assert row[0].tzinfo == zoneinfo.ZoneInfo("UTC")
+    assert row[1].tzinfo == zoneinfo.ZoneInfo("UTC")
```
Comment (joe-clickhouse) [inline tests/integration_tests/test_timezones.py]: We'll need to replace with `if tzutil.is_utc_timezone(row[...].tzinfo):`

File: tests/integration_tests/test_timezones.py [hunk 12]
```diff
@@ -212,8 +217,8 @@ def test_tz_mode(param_client: Client, call):
             query_tz="UTC",
             tz_mode="aware",
         ).first_row
-        assert row[0].tzinfo == pytz.UTC
-        assert row[1].tzinfo == pytz.UTC
+        assert row[0].tzinfo == zoneinfo.ZoneInfo("UTC")
+        assert row[1].tzinfo == zoneinfo.ZoneInfo("UTC")
```
Comment (joe-clickhouse) [inline tests/integration_tests/test_timezones.py]: We'll need to replace with `if tzutil.is_utc_timezone(row[...].tzinfo):`

**Review by joe-clickhouse — COMMENTED** [review]
Great work so far! Left some specific comments in the code.

File: clickhouse_connect/driver/tzutil.py [hunk 13]
```diff
@@ -13,50 +17,54 @@
 # timezone, but if someone insists on using the local timezone we will try to convert.  The problem is we
 # never have anything but an epoch timestamp returned from ClickHouse, so attempts to convert times when the
 # local timezone is "DST" aware (like 'CEST' vs 'CET') will be wrong approximately half the time
-local_tz: pytz.timezone
+local_tz: zoneinfo.ZoneInfo
```
Comment (joe-clickhouse) [inline clickhouse_connect/driver/tzutil.py]: `local_tz` should be type `tzinfo`.

File: clickhouse_connect/driver/tzutil.py [hunk 14]
```diff
@@ -13,50 +17,54 @@
...
 local_tz_dst_safe: bool = False
 
 # Timezone names that are equivalent to UTC
 UTC_EQUIVALENTS = ("UTC", "Etc/UTC", "GMT", "Universal", "GMT-0", "Zulu", "Greenwich", "UCT")
 
 
-def normalize_timezone(timezone: pytz.timezone) -> tuple[pytz.timezone, bool]:
+def normalize_timezone(timezone: zoneinfo.ZoneInfo) -> tuple[zoneinfo.ZoneInfo, bool]:
```
Comment (joe-clickhouse) [inline clickhouse_connect/driver/tzutil.py]: Similarly the types in and out should be `tzinfo`.

File: clickhouse_connect/driver/tzutil.py [hunk 15]
```diff
@@ -13,50 +17,54 @@
...
 def utcfromtimestamp(ts: float) -> datetime:
-    return datetime.fromtimestamp(ts, tz=pytz.UTC).replace(tzinfo=None)
+    return datetime.fromtimestamp(ts, tz=zoneinfo.ZoneInfo("UTC")).replace(tzinfo=None)
 
 
-try:
-    local_tz = pytz.timezone(os.environ.get("TZ", ""))
-except pytz.UnknownTimeZoneError:
-    local_tz = datetime.now().astimezone().tzinfo
+def local_tz() -> tzinfo:
```
Comment (joe-clickhouse) [inline clickhouse_connect/driver/tzutil.py]: Let's rename this helper to something like `_detect_local_tz()` so it doesn't clash with the module-level `local_tz`.

**Review by joe-clickhouse — COMMENTED** [review]
Added a few mote

**Comment by joe-clickhouse** [comment]
Hi @varoonp123 I really hope you don't mind, but I went ahead and just pushed a commit that included all the changes we discussed and I left comments for. I really trying to get a 1.0.0rc out ASAP and if this change doesn't make it in before that, it'd likely have to wait until 2.0. So thanks again for doing almost all the work here and providing the impetus to actually get this done! It's really appreciated.

**Comment by varoonp123** [comment]
@joe-clickhouse not at all. I would appreciate if we could merge tomorrow. Ill have some time tonight after work to take 1 more look/sanity check at this.

**Comment by joe-clickhouse** [comment]
@varoonp123 yes, of course! Sounds good. After you review it, please let me know if you have any questions/concerns/comments etc.

**Comment by varoonp123** [comment]
Ok it looks good to merge. Thanks for the help @joe-clickhouse !

File: clickhouse_connect/driver/tzutil.py [hunk 16]
```diff
@@ -13,50 +12,105 @@
...
-    if timezone.tzname(None) in pytz.common_timezones:
-        return timezone, True
+    if tz_key in zoneinfo.available_timezones():
+        return tz, True
 
     if tzlocal is not None:  # Maybe we can use the tzlocal module to get a safe timezone
         local_name = tzlocal.get_localzone_name()
-        if local_name in pytz.common_timezones:
-            return pytz.timezone(local_name), True
+        if local_name in zoneinfo.available_timezones():
```
[the remark this answers is not in the document]
Comment (joe-clickhouse) [reply clickhouse_connect/driver/tzutil.py:81]: Won't fix.

File: clickhouse_connect/driver/tzutil.py [hunk 17]
```diff
@@ -13,50 +12,105 @@
...
+    no system zoneinfo and the tzdata package is not installed.
+    """
+    if tz_name in UTC_EQUIVALENTS:
+        return timezone.utc
+    try:
+        return zoneinfo.ZoneInfo(tz_name)
+    except ValueError as ex:
+        # ZoneInfo raises ValueError for empty strings, absolute paths, and non-normalized
+        # keys; funnel those into ZoneInfoNotFoundError so callers only need one except clause.
+        raise zoneinfo.ZoneInfoNotFoundError(str(ex)) from ex
```
[the remark this answers is not in the document]
Comment (joe-clickhouse) [reply clickhouse_connect/driver/tzutil.py:63]: Fine as is, imo.
